### PR TITLE
Generate docker-compose.yml file in PHP

### PR DESCRIPTION
**Pull request by joehoyle**
We are reaching the limitations of what we can do with a static docker-compose file. There's no way to optionally disable services for example, or provide configuration for the containers from the `composer.json` or CLI flags. This moves the definition of the docker-compose file into PHP, and will dump it out to a docker-compose.yml file for use in `docker-compose`.

### CODE REVIEW

**Comment by roborourke** [comment]
We discussed in slack briefly but adding my high level thoughts here for the records:

I'm +1 on this approach, gives a huge amount more flexibility to tailor the environment through config and should simplify passing environment variables around as there's only one level we need to read them at this way.

**Comment by nathanielks** [comment]
This sounds great 👍

**Comment by joehoyle** [comment]
@roborourke Ok I updated this with the latest changes in master, and I think we should go with it before it gets stale again. I've ro-organized the generated file, and also committed an example build over the old docker-composer.yml, so we can look at the diff here to see what it's going to generate differently. A second pair of eyes on the generated docker-compose.yml would be great!

**Comment by joehoyle** [comment]
With the latest change, you can also put ```        "local-server": {
          "elasticsearch": false,
          "cavalcade": false,
          "analytics": false
        },```

In your project's config, to have a much slimmer local env. We _could_ have these instead inherit from the main modules that use them, so if you set `modules.analytics.enabled = false` we could default also to not running those. I feel like this might be a better way to go (or support it as the default, and allow the `modules.local-server.analytics` as a higher priority override.

**Comment by nathanielks** [comment]
eg load the config that exists in the project and modify just the fields that need modification?

**Comment by joehoyle** [comment]
> eg load the config that exists in the project and modify just the fields that need modification?

I don't think long-term this will be more maintainable. I think `php => yaml` is going to be better than `yaml => php => yaml`. Also, there are subtleties in output formatting that won't make it _exactly_ the same as the original docker-compose.yml, so I don't think it helps a lot in being able to cleanly diff/discern exactly what was changed.

**Comment by roborourke** [comment]
> With the latest change, you can also put
> 
> ```
>           "elasticsearch": false,
>           "cavalcade": false,
>           "analytics": false
> },
> ```
> 
> In your project's config, to have a much slimmer local env. We _could_ have these instead inherit from the main modules that use them, so if you set `modules.analytics.enabled = false` we could default also to not running those. I feel like this might be a better way to go (or support it as the default, and allow the `modules.local-server.analytics` as a higher priority override.

That does make sense but it'll be somewhat complicated I think as we'd need to get the merged config to ensure the local env overrides are applied too. I'm not sure we can do that easily in composer commands as we've had to just read directly from `composer.json` to do these in the past but I might be wrong if the main Altis core namespace is autoloaded.

File: docker/docker-compose.yml
Comment (roborourke) [inline docker/docker-compose.yml:118]: This value seems to have changed compared to the previous

File: docker/docker-compose.yml
Comment (roborourke) [inline docker/docker-compose.yml:245]: Missing the `--overwrite` flag here

File: docker/docker-compose.yml
Comment (roborourke) [inline docker/docker-compose.yml:284]: Need to set this to `unless-stopped`

File: inc/composer/class-docker-compose-generator.php
Comment (roborourke) [inline inc/composer/class-docker-compose-generator.php:117]:
```suggestion
					'restart' => 'unless-stopped',
```

File: inc/composer/class-docker-compose-generator.php
Comment (roborourke) [inline inc/composer/class-docker-compose-generator.php:148]:
```suggestion
			],
```

File: inc/composer/class-docker-compose-generator.php
Comment (roborourke) [inline inc/composer/class-docker-compose-generator.php:324]:
```suggestion
				'entrypoint' => "/bin/sh -c \"mc mb -p local/s3-{$this->project_name} && mc policy set public local/s3-{$this->project_name} && mc mirror --watch --overwrite local/s3-{$this->project_name} /content\"",
```

File: inc/composer/class-docker-compose-generator.php
Comment (roborourke) [inline inc/composer/class-docker-compose-generator.php:389]:
```suggestion
				'restart' => 'unless-stopped',
```

File: inc/composer/class-docker-compose-generator.php
Comment (roborourke) [inline inc/composer/class-docker-compose-generator.php:406]:
```suggestion
				'restart' => 'unless-stopped',
```

File: inc/composer/class-docker-compose-generator.php
Comment (roborourke) [inline inc/composer/class-docker-compose-generator.php:459]:
```suggestion
			$this->get_service_mailhog()
```

File: inc/composer/class-docker-compose-generator.php
Comment (roborourke) [inline inc/composer/class-docker-compose-generator.php:461]: Analytics depends on elasticsearch so that should be factored in here
```suggestion
		if ( $this->get_config()['analytics'] && $this->get_config()['elasticsearch'] ) {
```

File: inc/composer/class-docker-compose-generator.php
Comment (roborourke) [inline inc/composer/class-docker-compose-generator.php:464]: Better to allow kibana to be required or not, eg. we don't need it in Travis or CI envs, at least not by default.
```suggestion
		if ( $this->get_config()['kibana'] && $this->get_config()['elasticsearch'] ) {
```

File: inc/composer/class-docker-compose-generator.php
Comment (roborourke) [inline inc/composer/class-docker-compose-generator.php:506]:
```suggestion
			'elasticsearch' => true,
			'kibana' => true,
```

**Review by roborourke — CHANGES_REQUESTED** [review]
I've added in the missing bits / out of date bits as suggestions but this looks good on the whole. Couple of other things:

- Is `db-data:` equivalent to `db-data: null` eg. in the volumes section?
- The docker compose YAML should be output during the `start` command. Right now you have to dump the YAML then start the server
   - I've generally used the vendor directory as a place to dump generated files like this as it's guaranteed to be git ignored
- The class file code style needs to reflect the rest of what we do, eg. doc blocks and some more inline comments might be useful where there are `if` statements used in the method that deals with config

**Comment by nathanielks** [comment]
> so I don't think it helps a lot in being able to cleanly diff/discern exactly what was changed.

I recognize this is a matter of preference, but I do think using a template file would simplify development and provide a far cleaner diff (I don't think we'd even need a yaml parser, necessarily). In essence, I would advocate for what we're doing in our Terraform configuration. We have template files that posses placeholder tokens that get replaced by the [`template_file`](https://registry.terraform.io/providers/hashicorp/template/latest/docs/data-sources/file) data source. I see a lot of what is effectively boilerplate in the current array syntax, and for someone whose hands are injured, I try to reduce boilerplate where I can. Using a template file would reduce the effort to a `file_get_contents()`, `sprintf()`, and `file_put_contents()` 🙂

**Comment by nathanielks** [comment]
That said, I definitely won't block this PR but I wanted to hopefully make our lives easier!

**Comment by joehoyle** [comment]
>  Using a template file would reduce the effort to file_get_contents(), sprintf(), and file_put_contents() 🙂

The intention of this PR is to open up a lot more flexibility than what can be done this way. The conditional loading of services (which also can change the healthecks) is one example, but I believe taking this approach will open up more possibilities too.

**Comment by nathanielks** [comment]
By all means, carry on!

**Comment by joehoyle** [comment]
@roborourke should we get this into b2, or go with v6 at this point?

**Comment by roborourke** [comment]
I think v6 as it's a fundamental change and we should keep to the feature / code freeze to keep things manageable. We would also want to add things that integrate and make use of it at the same time ideally such as removing unneeded services in CI.

**Review by roborourke — APPROVED** [review]

**Pull request merged**
